### PR TITLE
[schemaregistry]ProtobufNative Schema Support

### LINKED ISSUE #7642: Protobuf Schema support AutoConsumeSchema and Presto table
**Is your enhancement request related to a problem? Please describe.**
1. Now , `ProtobufSchema` create and persistence dependency Java implementation class of GeneratedMessageV3 , create action code trace is  [ProtobufSchema.of(Class<T> pojo) ](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufSchema.java#L102) ->[org.apache.avro.protobuf.ProtobufData.getSchema(Class c)](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufSchema.java#L64), but avro-protobuf lib i haven't find method to reflect class by Descriptor，so [ProtobufSchema.loadReader()](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufSchema.java#L98) is don't support , and [AutoConsumeSchema.generateSchema()](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AutoConsumeSchema.java#L149) haven't support PROTOBUF。
3. I am working for  Presto pulsar table , but pulsar current protobuf comsumer need [Java implementation class of GeneratedMessageV3 ](https://github.com/apache/pulsar/blob/master/pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleTypedProducerConsumerTest.java), Presto is a Long-Running server, so top server then update class is unfriendly。
4.  Regist ProtobufSchema with Java implementation class of GeneratedMessageV3 is unfriendly for other languages instead of JAVA User :  [`SimpleTypedProducerConsumerTest`](https://github.com/apache/pulsar/blob/master/pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleTypedProducerConsumerTest.java)

------

**Describe the solution you'd like**

**AutoConsumeSchema** support Protobuf , Presto table work with Protobuf schema register

1. ProtobufSchema regist use protoc **descriptor_set_out** file,  persistence this single file content to Schema Registry，.desc conten  is language-independent and readable .
e.g. ``protoc --include_imports --descriptor_set_out=<filename>.desc <filename>.proto <filename>.proto <filename>.proto...``
2.  consumer support Protobuf **AutoConsumeSchema by DynamicMessage** ,e.g.
``` 
BufferedInputStream dscFile = new BufferedInputStream(new FileInputStream(new File ("desc file path")).getPath())
DescriptorProtos.FileDescriptorSet fileDescriptorSet = DescriptorProtos.FileDescriptorSet.parseFrom(dscFile);
```
- **single_message** example:

``DynamicMessage.parseFrom(fileDescriptor.getMessageTypes().get(0), outputStream.toByteArray());``

- **multiple_message** example :
``
DynamicMessage.parseFrom(fileDescriptor.findMessageTypeByName(${MessageName}), outputStream.toByteArray());``
MessageName  persistence with SchemaInfo.properties when create。
3. Then Presto ProtobufSchemaHandler can extract  Field(etc. Value\List\Map ) from DynamicMessage 

------

**Additional context**
If this proposal is approved，I glad to make a more detailed description or PR 。

**tag** : `Client \Schema Registry  with Protobuffer` , `Presto`

### LINKED ISSUE #7674: [Java Client] Schema Registry enhancement for protobuf schema
We are trying to encode/decode protobuf messages from/to json and we don't have the .class files of protobuf message, there is no way for us to achieve that, because, as my knowledge, we have to have a descriptor file to be able to encode/decode protobuf messages, but the schema registry only store json schema of protobuf message.

**Suggested Solution**
Whenever a producer upload a new schema to a topic, it should also attach the protobuf descriptor file content in schema properties. With the descriptor is attached in schema registry, the client will be able to encode/decode Protobuf binary data correctly

### PR DESCRIPTION
**Pull request by hnail**
Fixes #7642
Fixes #7674 

### Motivation
Protobuf  Native Schema Support :  [PIP-Add ProtobufNative Schema Support ](https://docs.google.com/document/d/1XR_MNOuSXyig-CKsdVhr6IXvFwziBRdSoS3oEUiLFe8/edit?usp=sharing)

This PR proposes to import a new protobuf-v3 schema based on protobuf native `Descriptor`, Current  `ProtobufSchema` is based on AVRO schema,  this Causes the following restrictions :

- Current [ProtobufSchema](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufSchema.java) based on avro-protobuf only works when JVM classloader can load `Java implementation class of GeneratedMessageV3` . this is unfriendly for long-running server ( e.g. Presto ), restart server for update class is unfriendly.
- Describe `protobuf` schema by `avro` schema will cause losses of information, so based current [ProtobufSchema](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufSchema.java) can’t support `AutoConsume` by `DynamicMessage`.
- The amount of support avro language is less than support protobuf language.

In consideration of backward compatibility , we add a new schema type  `SchemaType.PROTOBUF_NATIVE` base on protobuf-v3 native Descriptor instead of modify SchemaType.PROTOBUF, aim to support `GenericProtobufNativeSchema` and `AutoConsumeSchema` for PROTOBUF.

### Modifications
Describe in [PIP-Add ProtobufNative Schema Support ](https://docs.google.com/document/d/1XR_MNOuSXyig-CKsdVhr6IXvFwziBRdSoS3oEUiLFe8/edit?usp=sharing)

----

##### Create ProtobufNative schema Example
```
SchemaDefinition def =  SchemaDefinition.<PBMessage>builder()
                .withAlwaysAllowNull(true)
                .withPojo(PBMessage.class)
                .build();
Schema schema = Schema.PROTOBUFNATIVE(def);
admin.schemas().createSchema(topic, schema.getSchemaInfo());
```
##### Class-Based schema consumer Example
```
Consumer<PBMessage> consumer = client.newConsumer(Schema.PROTOBUFNATIVE(PBMessage.class)).topic(topic)
                .subscriptionName("my-subscription-name").subscriptionInitialPosition(SubscriptionInitialPosition.Earliest).subscribe();
        while (!consumer.hasReachedEndOfTopic()){
            Message<PBMessage> msg = consumer.receive();
            PBMessage pBMessage =  msg.getValue();
            System.out.println("getPlatform : "+ pBMessage.getPlatform());
        }
```
##### ProtobufNative AUTO_CONSUME  Example
```
 Consumer<GenericRecord> consumer = client.newConsumer(Schema.AUTO_CONSUME()).topic(topic)
                .subscriptionName("my-subscription-nameaaa").subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
        .subscribe();
        while (!consumer.hasReachedEndOfTopic()){
            Message<GenericRecord> msg = consumer.receive() ;
           // System.out.println("GenericRecord :"+msg.getValue().toString());
            GenericRecord genericRecord =  msg.getValue();
            System.out.println(genericRecord.getField("platform"));
            GenericProtobufNativeRecord genericProtobufNativeRecord = (GenericProtobufNativeRecord)genericRecord;
            DynamicMessage dynamicMessage = genericProtobufNativeRecord.getProtobufRecord();         
            String platform = dynamicMessage.getField(dynamicMessage.getDescriptorForType().findFieldByName("platform")).toString();
            System.out.println("platform : "+ platform);
        }
```

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: ( **yes** )
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? ([PIP-Add ProtobufNative Schema Support ](https://docs.google.com/document/d/1XR_MNOuSXyig-CKsdVhr6IXvFwziBRdSoS3oEUiLFe8/edit?usp=sharing))

### CODE REVIEW

**Comment by sijie** [comment]
@hnail Can you send the PIP of adding native protobuf support to dev@ mailing list? So the community can review the PIP and we can add the PIP to Pulsar's wiki page.

**Comment by hnail** [comment]
> @hnail Can you send the PIP of adding native protobuf support to dev@ mailing list? So the community can review the PIP and we can add the PIP to Pulsar's wiki page.

send <[DISCUSS] PIP-Add ProtobufNative Schema Support> to dev@pulsar.apache.org   done

**Comment by hnail** [comment]
@sijie @codelipenghui Has there been progresses on this review ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/ProtobufNativeSchemaCompatibilityCheck.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Override
+    public void checkCompatible(Iterable<SchemaData> from, SchemaData to, SchemaCompatibilityStrategy strategy) throws IncompatibleSchemaException {
+        for (SchemaData schemaData : from) {
+            checkCompatible(schemaData, to, strategy);
+        }
+    }
+
+    private void CheckRootRootMessageChange(Descriptor fromDescriptor, Descriptor toDescriptor, SchemaCompatibilityStrategy strategy) throws IncompatibleSchemaException {
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/ProtobufNativeSchemaCompatibilityCheck.java:66]: why there is "RootRoot" in the method name `CheckRootRootMessageChange`?

Also, we need to make the first character low case.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/validator/ProtobufNativeSchemaDataValidator.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+public class ProtobufNativeSchemaDataValidator implements SchemaDataValidator {
+
+    @Override
+    public void validate(SchemaData schemaData) throws InvalidSchemaDataException {
+        Descriptors.Descriptor descriptor;
+        try {
+            descriptor = ProtobufNativeSchemaUtils.deserialize(schemaData.getData());
+        } catch (Exception e) {
+            throw new InvalidSchemaDataException("deserialize  ProtobufNative Schema failed", e);
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/validator/ProtobufNativeSchemaDataValidator.java:34]:
```suggestion
            throw new InvalidSchemaDataException("deserialize ProtobufNative Schema failed", e);
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/validator/ProtobufNativeSchemaDataValidator.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        if (descriptor == null) {
+            throw new InvalidSchemaDataException("protobuf root message Descriptor is null , please recheck rootMessageTypeName or rootFileDescriptorName conf. ");
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/validator/ProtobufNativeSchemaDataValidator.java:37]:
```suggestion
            throw new InvalidSchemaDataException("protobuf root message descriptor is null , please recheck rootMessageTypeName or rootFileDescriptorName conf. ");
```

**Review by sijie — APPROVED** [review]

**Comment by sijie** [comment]
@hnail overall the change looks really good. I made a few comments. PTAL.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaUtils.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import org.apache.pulsar.common.protocol.schema.ProtobufNativeSchemaData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.protobuf.DescriptorProtos.FileDescriptorProto;
+import static com.google.protobuf.DescriptorProtos.FileDescriptorSet;
+import static com.google.protobuf.Descriptors.*;
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaUtils.java:33]: Avoid use import.* here

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeSchema.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl.schema.generic;
+
+import com.google.protobuf.Descriptors;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.client.api.schema.*;
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericProtobufNativeSchema.java:23]: Avoid use import.* here

**Review by codelipenghui — APPROVED** [review]
Looks good to me, just left some minor comments.

**Comment by codelipenghui** [comment]
@congbobo184 Could you please help review this PR?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/ProtobufNativeSchemaCompatibilityCheck.java (hunk 1 above)
Comment (hnail) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/ProtobufNativeSchemaCompatibilityCheck.java:66]: @sijie thanks for the reminders , It was my carelessness, this shoud be `CheckRootMessageChange ` , I will recheck my code details carefully.

Comment (hnail) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/ProtobufNativeSchemaCompatibilityCheck.java:66]: @sijie thanks for the reminders , it due to my carelessness . this shoud be `CheckRootMessageChange `, I may need recheck my code details carefully

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaUtils.java (hunk 4 above)
Comment (hnail) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaUtils.java]: @codelipenghui thanks for review , fixed as Suggestion。

**Comment by fjod** [comment]
I have a question about FileDescriptorSet field. In https://github.com/apache/pulsar/blob/c01b1eeda3221bdbf863bf0f3f8373e93d90adef/pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaTest.java  file there is an example test class. It is generated from Test.proto and ExternalMessage.proto files. The problem is that, no matter what I try with protoc , I can not get same FileDescriptorSet  content. It's always slightly different. I tried csharp, java, cpp output, with or without --descriptor_set_out - I always get different byte array (and base64 string). And I can not create producer/consumer on apache pulsar client with FileDescriptorSet  generated by me, I get com.google.protobuf.InvalidProtocolBufferException.invalidEndTag exception which as far as I googled tells that data is corrupted. When I try to create producer with FileDescriptorSet   data from test, it works. What am I doing wrong while generating descriptors using protoc ?

**Comment by hnail** [comment]
> I have a question about FileDescriptorSet field. In https://github.com/apache/pulsar/blob/c01b1eeda3221bdbf863bf0f3f8373e93d90adef/pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaTest.java file there is an example test class. It is generated from Test.proto and ExternalMessage.proto files. The problem is that, no matter what I try with protoc , I can not get same FileDescriptorSet content. It's always slightly different. I tried csharp, java, cpp output, with or without --descriptor_set_out - I always get different byte array (and base64 string). And I can not create producer/consumer on apache pulsar client with FileDescriptorSet generated by me, I get com.google.protobuf.InvalidProtocolBufferException.invalidEndTag exception which as far as I googled tells that data is corrupted. When I try to create producer with FileDescriptorSet data from test, it works. What am I doing wrong while generating descriptors using protoc ?

hello , happy for your take notice of this pull request .
I test as your description with the two following test case :

- FileDescriptorSet build by JAVA Proto Class  , same as [ProtobufNativeSchemaUtils#serialize](https://github.com/hnail/pulsar/blob/2597f2c6287783ba285737a28cb39cf3e058aa37/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaUtils.java#L53)

```
// buid FileDescriptorSet byte[] by Class  
byte[] fileDescriptorSetBytesByClass = DescriptorProtos.FileDescriptorSet.newBuilder()
                .addFile(Service.ServiceRequest.getDescriptor().getFile().toProto())
                .build().toByteArray();  
DescriptorProtos.FileDescriptorSet fileDescriptorSetByClass = DescriptorProtos.FileDescriptorSet
                .parseFrom(fileDescriptorSetBytesByClass);
// print FileDescriptorSet string which buid by Class
System.out.println(new String(fileDescriptorSetByClass.toBuilder().build().toByteArray(),"utf-8"));
```
- FileDescriptorSet build by protoc command , same as your description:
```
// buid FileDescriptorSet byte[] by 'protoc --include_imports --descriptor_set_out Request.desc Request.proto
byte[] fileDescriptorSetBytesByProtoC = FileUtils.readFileToByteArray(new File("/Users/wangguowei/source" +
                "/pulsar/dev/pulsar_test/src/main/resources/pulsar/Request.desc"));

DescriptorProtos.FileDescriptorSet fileDescriptorSetByProtoC = DescriptorProtos.FileDescriptorSet
                .parseFrom(fileDescriptorSetBytesByProtoC);

// print FileDescriptorSet string which buid by ProtoC
System.out.println(new String(fileDescriptorSetByClass.toBuilder().build().toByteArray(),"utf-8"));
```
The two realize above works  for me : 
- The serialized FileDescriptorSet byte[] is slightly different as your description, I think maybe java `proto class` lack of information compare with `proto file ` when compile java class .
- Deserialize  FileDescriptorSet is by `DescriptorProtos.FileDescriptorSet.parseFrom(bytes[])` in  two way above [ProtobufNativeSchemaUtils#deserialize](https://github.com/hnail/pulsar/blob/2597f2c6287783ba285737a28cb39cf3e058aa37/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaUtils.java#L90) . so , even though byte[] is slightly different , but at the same of working .
- The current realize is reference google-doc :  [protobuf v3 Self-describing Messages](https://developers.google.com/protocol-buffers/docs/techniques#self-description)

--- 
so , I think the reason is `new ObjectMapper().writeValueAsBytes(schemaData);` which serialize byte[] to String , may be is 'UTF-8' or 'Base64' ?

**Comment by fjod** [comment]
I found the culprit here, I deleted important information from source .proto file like **java_package**  and **java_outer_classname**, and protoc was generating different info which apache pulsar was not able to use.

**Comment by hnail** [comment]
> I found the culprit here, I deleted important information from source .proto file like **java_package** and **java_outer_classname**, and protoc was generating different info which apache pulsar was not able to use.

I recommend  you can use  JAVA API to create Schema like this : 
```
SchemaDefinition def =  SchemaDefinition.<ProtoMsg>builder()
                .withPojo(ProtoMsg.class)
                .build();
Schema schema = Schema.PROTOBUF_NATIVE(def);
admin.schemas().createSchema(topic, schema.getSchemaInfo());
```
The **java_package** and **java_outer_classname** restrictions which you mentioned may need check in other language client when create schema by API , be appreciate if your have time to help pull a issue to describe  the problem for future optimization .

**Comment by fjod** [comment]
> The **java_package** and **java_outer_classname** restrictions which you mentioned may need check in other language client when create schema by POJO API , be appreciate if your have time to help pull a issue to describe the problem for future optimization .

I created issue here about csharp/java mismatch in generating Descriptors

https://github.com/protocolbuffers/protobuf/issues/8425

**Comment by fjod** [comment]
@hnail I think some bug in deserialize method of ProtobufNativeSchemaUtils. I generated descriptor using probuf-net library, here it is **CgpUZXN0LnByb3RvEgVwcm90byImCgpYWFhNZXNzYWdlEgsKA2ZvbxgBIAEoCRILCgNiYXIYAiABKAFCD6oCDG5hdGl2ZUNoZWNrMmIGcHJvdG8z** .  If you check it on https://protogen.marcgravell.com/decode , it decodes fine. Here is json file with some additional fields:

{"fileDescriptorSet":"ClMKClRlc3QucHJvdG8SC25hdGl2ZUNoZWNrIjAKClhYWE1lc3NhZ2USEAoDZm9vGAEgASgJUgNmb28SEAoDYmFyGAIgASgBUgNiYXJiBnByb3RvMw==","rootMessageTypeName":"proto.XXXMessage","rootFileDescriptorName":"Test.proto"}

When I try to send it to apache-pulsar server, I get null-pointer exception on line 104 at  https://github.com/apache/pulsar/blob/2a1828ca08f096a94b696737e93e84a2bc60a8cc/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaUtils.java#L104

I created sample java app copy-pasting code from ProtobufNativeSchemaUtils. Here is simpler version of deserializer:

```
         File file;
         Descriptors.Descriptor descriptor;
         file = new File("D:\\serialized");
         byte[] fileContent = Files.readAllBytes(file.toPath());
         ProtobufNativeSchemaData schemaData = new ObjectMapper().readValue(fileContent, ProtobufNativeSchemaData.class);
         FileDescriptorSet fileDescriptorSet = FileDescriptorSet.parseFrom(schemaData.fileDescriptorSet);   
        Map<String, FileDescriptorProto> fileDescriptorProtoCache = new HashMap<>();
        Map<String, Descriptors.FileDescriptor> fileDescriptorCache = new HashMap<>();
        fileDescriptorSet.getFileList().forEach(fileDescriptorProto -> fileDescriptorProtoCache.put(fileDescriptorProto.getName(), fileDescriptorProto));        
        FileDescriptorProto rootFileDescriptorProto = fileDescriptorProtoCache.get("Test.proto");
        deserializeFileDescriptor(rootFileDescriptorProto, fileDescriptorCache, fileDescriptorProtoCache);
        Descriptors.FileDescriptor fileDescriptor = fileDescriptorCache.get("Test.proto");      
        System.out.print(fileDescriptor.toProto());
```

output is:

```
name: "Test.proto"
package: "nativeCheck"
message_type {
  name: "XXXMessage"
  field {
    name: "foo"
    number: 1
    label: LABEL_OPTIONAL
    type: TYPE_STRING
    json_name: "foo"
  }
  field {
    name: "bar"
    number: 2
    label: LABEL_OPTIONAL
    type: TYPE_DOUBLE
    json_name: "bar"
  }
}
syntax: "proto3"
```

**So Descriptor is fine**! But if I try to use code from the repo:

```
        Descriptors.Descriptor descriptor2;
        Descriptors.FileDescriptor fileDescriptor2 = fileDescriptorCache.get(schemaData.rootMessageTypeName);
        String package1 = fileDescriptor2.getPackage();
                //trim package
        String[] paths = StringUtils.removeFirst(schemaData.rootMessageTypeName, package1).replaceFirst("\\.", "").split("\\.");
        //extract root message
        descriptor2 = fileDescriptor2.findMessageTypeByName(paths[0]);
        //extract nested message
        for (int i = 1; i < paths.length; i++) {
            descriptor2 = descriptor2.findNestedTypeByName(paths[i]);
        }
        System.out.print(descriptor2.toProto());
```

I get nullpointer exception on line with getPackage(); - similar to apache-pulsar exception. 
image
I guess our generated descriptor does not contain such "package" and fails therefore? It would be great if we can use descriptor from our client too ;)

**Comment by hnail** [comment]
@fjod  Hi , Your `rootMessageTypeName` is incorrect , set to `nativeCheck.XXXMessage` , rootMessageTypeName must be **full path** , `nativeCheck` is your package path .

Simpler Test code :
```
String schemaDataBytes = "{\"fileDescriptorSet" +
                "\":\"ClMKClRlc3QucHJvdG8SC25hdGl2ZUNoZWNrIjAKClhYWE1lc3NhZ2USEAoDZm9vGAEgASgJUgNmb28SEAoDYmFyGAIgASgBUgNiYXJiBnByb3RvMw==\",\"rootMessageTypeName\":\"nativeCheck.XXXMessage\",\"rootFileDescriptorName\":\"Test.proto\"}";
        
Descriptors.Descriptor descriptor = ProtobufNativeSchemaUtils.deserialize(schemaDataBytes.getBytes());

System.out.println(descriptor);
```
It is works fine .

**Pull request merged**